pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -0,0 +1,409 @@
+		err := c.vSphereConnection.Logout(ctx)
+		if err != nil {
+			klog.Errorf("%s: error closing connection to vCenter API: %v", c.name, err)
+		}
+	}()
+
+	// if there was an OCP error we should degrade the cluster or if we previously created storageClass
+	// but we can't connect to vcenter now, we should also degrade the cluster
+	err, degradedOrBlockedFromUpgrades := c.blockUpgradeOrDegradeCluster(ctx, connectionResult)
+	if err != nil {
+		return err
+	}
+	if degradedOrBlockedFromUpgrades {
+		return nil
+	}
+
+	// sync storage class
+	// TODO: There are two questions unanswered here:
+	//	1. Should we sync/create storageclass even if cluster check fails?
+	//	2. Should cluster check not be peformed when storage class sync fails?
+	storageClassSyncResult := c.storageClassController.Sync(ctx, c.vSphereConnection)
+
+	err, degradedOrBlockedFromUpgrades = c.blockUpgradeOrDegradeCluster(ctx, storageClassSyncResult)
+	if err != nil {
+		return err
+	}
+	if degradedOrBlockedFromUpgrades {
+		return nil
+	}
+
+	delay, result, checkRan := c.runClusterCheck(ctx)
+
+	// if checks did not run
+	if !checkRan {
+		return nil
+	}
+	queue := syncContext.Queue()
+	queueKey := syncContext.QueueKey()
+
+	klog.V(2).Infof("Scheduled the next check in %s", delay)
+	time.AfterFunc(delay, func() {
+		queue.Add(queueKey)
+	})
+
+	err, degradedOrBlockedFromUpgrades = c.blockUpgradeOrDegradeCluster(ctx, result)
+	if err != nil {
+		return err
+	}
+
+	// if operand was not started previously and block upgrade is false and clusterdegrade is also false
+	// then and only then we should start CSI driver operator
+	if !c.operandControllerStarted && !degradedOrBlockedFromUpgrades {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+	return nil
+}
+
+func (c *VSphereController) blockUpgradeOrDegradeCluster(ctx context.Context, result checks.ClusterCheckResult) (error, bool) {
+	var clusterCondition string
+	blockUpgrade := c.shouldDegradeCluster(result)
+	if blockUpgrade {
+		clusterCondition = "degraded"
+		installErrorMetric.WithLabelValues(string(result.CheckStatus), clusterCondition).Set(1)
+		return result.CheckError, true
+	}
+	if result.BlockUpgrade {
+		clusterCondition = "upgrade_blocked"
+		installErrorMetric.WithLabelValues(string(result.CheckStatus), clusterCondition).Set(1)
+		updateError := c.updateConditions(ctx, c.name, result)
+		return updateError, true
+	}
+	return nil, false
+}
+
+func (c *VSphereController) shouldDegradeCluster(result checks.ClusterCheckResult) bool {
+	if result.ClusterDegrade {
+		return true
+	}
+
+	if result.BlockUpgrade {
+		// if we were supposed to only block upgrades but for some reason we have created CSIDriver and storageclass
+		// in past and future syncing fails then we should degrade the cluster
+		csiDriver, err := c.csiDriverLister.Get(utils.VSphereDriverName)
+
+		// if we can't fetch CSI driver lets not degrade the cluster, if cluster indeed needs degrading we will catch in
+		// next sync
```

Comment:
The next sync can be quite far away, in 20 minutes in the worst case. IMO degrading the cluster is appropriate here (ie. return err)

Reply:
changed. Also improved the check to check for `CSIDriver` object installed by OCP.